lib/tree.rb

```diff
@@ -1,64 +1,179 @@
 class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
-
-   def initialize(key, val)
+  
+  def initialize(key, value)
     @key = key
-    @value = val
+    @value = value
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
   def initialize
     @root = nil
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: Olog(n) - only need to look at half the tree
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+    
+    if @root.nil?
+      @root = new_node
+    else
+      add_helper(@root, new_node)
+    end
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def add_helper(current, new_node)
+    return new_node if current.nil?
+    
+    if new_node.key <= current.key
+      current.left = add_helper(current.left, new_node)
+    else
+      current.right = add_helper(current.right, new_node)
+    end
+    
+    return current
+  end
+  
+  # Time Complexity: Olog(n) - only need to look at half the tree
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    return nil if @root.nil?
+    return @root.value if @root.key == key
+    
+    current = @root
+    while current != nil
+      if key == current.key
+        return current.value
+      elsif key < current.key
+        current = current.left
+      elsif key > current.key
+        current = current.right
+      end
+    end
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(n) - n is the number of nodes in the tree
+  # Space Complexity: O(n) - my thought is that since we're making a list of nodes to return, we'll have that many nodes stored in the list, but since the stack doesn't ever hold something that's also already in the list, there would never be more than n nodes stored at one time..?  Let me know if that's on the right track.
+  # left, root, right (me btwn my children)
   def inorder
```

Comment:
Nice work doing this with loops and a stack.  I guess you **REALLY** don't like recursion.